Client/SMTP_client.cpp

```diff
@@ -0,0 +1,1006 @@
+{
+	SecureZeroMemory(m_receive_buffer.get(), DEFAULT_BUFFER_SIZE);
+}
+
+bool SMTPClientClass::set_server_choice(const std::string& s)
+{
+	if (s.size())
+	{		
+		m_smtp_address = s;
+		return true;
+	}	
+	else
+	{
+		return false;
+	}	
+}
+
+std::string SMTPErrorClass::GetErrorText() const
+{
+	switch (m_error_code)
+	{
+	case SMTPerrorEnum::SMTP_NO_ERROR:
+		return "No errors";
+	case SMTPerrorEnum::SSL_ERROR:
+		return "Error during SSL connection";
+	case SMTPerrorEnum::WSA_STARTUP_ERROR:
+		return "Unable to initialize Winsock2";
+	case SMTPerrorEnum::WSA_HOSTNAME_ERROR:
+		return "Couldn't open connection to the chosen host";
+	case SMTPerrorEnum::WSA_INVALID_SOCKET_ERROR:
+		return "Invalid Winsock2 socket";
+	case SMTPerrorEnum::WSA_SELECT_ERROR:
+		return "Function select() error";
+	case SMTPerrorEnum::WSA_SEND_ERROR:
+		return "Function send() error";
+	case SMTPerrorEnum::WSA_RECEIVE_ERROR:
+		return "Function recv() error";
+	case SMTPerrorEnum::SERVER_NOT_RESPONDING:
+		return "Server not responding";
+	case SMTPerrorEnum::SERVER_HELLO_ANSWER:
+		return "Server returned error after sending EHLO";
+	case SMTPerrorEnum::SERVER_AUTH_LOGIN_ANSWER:
+		return "Server returned error after sending AUTH LOGIN";
+	case SMTPerrorEnum::SERVER_AUTHORIZATION_FAILED:
+		return "Server failed to authorizate user";
+	case SMTPerrorEnum::SERVER_MAIL_FROM_ANSWER:
+		return "Server returned error after sending MAIL FROM:";
+	case SMTPerrorEnum::SERVER_RCPT_TO_ANSWER:
+		return "Server returned error after sending RCPT TO:";
+	case SMTPerrorEnum::SERVER_DATA_ANSWER:
+		return "Server returned error after sending DATA";
+	case SMTPerrorEnum::SERVER_LETTER_ANSWER:
+		return "Server returned error after sending letter body";
+	case SMTPerrorEnum::SERVER_QUIT_ANSWER:
+		return "Server returned error after sending QUIT, but your letter should be delivered";
+	case SMTPerrorEnum::SEND_MSG_EMPTY:
+		return "Sending message is empty";
+	case SMTPerrorEnum::LACK_OF_BUFFER_MEMORY:
+		return "System couldn't allocate enough memory for receive buffer";
+	case SMTPerrorEnum::CONNECTION_CLOSED:
+		return "Couldn't finished operation, because connection with server closed";
+	case SMTPerrorEnum::STRING_ARGUMENT_EMPTY:
+		return "Couldn't set value, because argument was empty string";
+	case SMTPerrorEnum::UNDEF_USER_LOGIN:
+		return "User login is undefined";
+	case SMTPerrorEnum::UNDEF_USER_PASSWORD:
+		return "User password is undefined";
+	case SMTPerrorEnum::SERVER_NOT_READY:
+		return "Server returned 'NOT READY' after connection was open";
+	case SMTPerrorEnum::UNDEF_SERVER_CHOICE:
+		return "User didn't define server to connect";
+	default:
+		return "Undefined error id";
+	}
+}
+
+SMTPSecureClientClass::SMTPSecureClientClass()
+{
+	try
+	{
+		m_receive_buffer.reset(new char[DEFAULT_BUFFER_SIZE]);
+	}
+	catch (...)
+	{
+		LOG_ERROR << "Server responded with error to EHLO";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPerrorEnum::LACK_OF_BUFFER_MEMORY);
```

Comment:
reset is a noexcept function. maybe no need to catch?
https://en.cppreference.com/w/cpp/memory/unique_ptr/reset

Reply:
I though that 'new' will throw if the system couldn't allocate memory, that we were asking.